cpp/src/arrow/compute/exec/filter_node.cc

```diff
@@ -98,12 +97,25 @@ class FilterNode : public ExecNode {
 
   void InputReceived(ExecNode* input, ExecBatch batch) override {
     DCHECK_EQ(input, inputs_[0]);
-
-    auto maybe_filtered = DoFilter(std::move(batch));
-    if (ErrorIfNotOk(maybe_filtered.status())) return;
-
-    maybe_filtered->guarantee = batch.guarantee;
-    outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+    if (finished_.is_finished()) {
+      return;
+    }
+    auto task = [this, batch]() {
+      auto maybe_filtered = DoFilter(std::move(batch));
+      if (ErrorIfNotOk(maybe_filtered.status())) return Status::OK();
+      maybe_filtered->guarantee = batch.guarantee;
+      outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+      return Status::OK();
+    };
+    if (this->has_executor()) {
+      DCHECK_OK(this->SubmitTask(task));
+    } else {
+      DCHECK_OK(task());
+    }
+    if (batch_count_.Increment()) {
+      task_group_.WaitForTasksToFinish().AddCallback(
+          [this](const Status& status) { this->finished_.MarkFinished(status); });
+    }
```

Comment:
Could this be part of SubmitTask so that this if/else loop doesn't have to belong in every node?